### ISSUE TITLE
sql: logic tests should enforce rowsort when necessary

### ISSUE THREAD

**Issue by justinj** [cockroachdb/cockroach#22261]
It's been a consistent source of test flakes that people do not specify `rowsort` on tests which need it, which is understandable because their tests pass locally.

We should figure out a good solution to this, for example, by linting that we rowsort tests which need it.

Moving this conversation out of slack, cc @benesch, @jordanlewis 

Jira issue: CRDB-5865

**Comment by knz** [comment]
related to #24223.

**Comment by knz** [comment]
I suppose we could use that as a first-order approximation. It probably has some false positives but at least it won't have false negatives (i.e. it can only lead us to using too many `rowsorts` but miss none)

**Comment by benesch** [comment]
It's worth pointing out though that rowsort on rows with embedded newlines is completely and entirely broken. (This is why I got stuck while working on the random shuffler that I posted in #24223.) I think a prerequisite to plumbing rowsort everywhere is introducing an alternate syntax to be used when rows contain newlines.

Here's an example of what I mean. The fact that logic test accepts this is pure dumb luck:

```
SHOW CREATE TABLE ok1
----
ok1  CREATE TABLE ok1 (
     a INT NOT NULL,
     b INT NOT NULL,
     c INT NULL,
     CONSTRAINT "primary" PRIMARY KEY (a ASC, b ASC),
     FAMILY "primary" (a, b, c)
) PARTITION BY LIST (a) (
   PARTITION p1 VALUES IN ((1)),
   PARTITION p2 VALUES IN ((2))
)
```

Granted, this example doesn't need rowsort because there's only one row, but I know there are logic tests that need rowsort and also have values with whitespace.

**Comment by knz** [comment]
@andy-kimball has implemented a solution to that problem in the `datadriven.go` code. I would suggest back-porting it to `logic.go`.

**Comment by andy-kimball** [comment]
`datadriven.go` is a bit different, since it's trying to create a general framework for "expected value" tests. The change I made there was to allow *blank* lines in the expected output. But that's different than the troubles encountered by @benesch, which stem from ambiguities that arise when the number of columns does not match the number of tokens for each result row. This happens when column values contain whitespace (whether newlines or spaces). Unfortunately, we've painted ourselves into a difficult corner with those tests, especially by the addition of `partialsort`. When I made changes to get the tests running with the cost-based optimizer, I discovered the following rules:

1. If no sorting is required, then we always compare by whitespace-delimited tokens, regardless of how they're split across rows. This algorithm works correctly, as long as no sorting options are used. No dumb luck required. For example:
```
col1     col2
-------------
foo bar  cat
baz      dog
```
In this case, it's ambiguous whether "bar" is part of the `col1` or `col2` (which matters to partial sort). Even `rowsort` can fail if results may have newlines, as Nikhil pointed out. But if no sort option is used, it turns out it doesn't matter - `logic.go` just compares a string of tokens: ("foo", "bar", "cat", "baz", "dog"). It doesn't matter how these values were originally split across rows/columns.

2. If we're running the `bigtest` suite, or if there is only one column in the result, we assume there is one result value per result line:
```
col
-------
foo bar
baz
```
That result has 2 values ("foo bar", "baz"), not 3 ("foo", "bar", "baz"), and can be used with sorting options. It's true this wouldn't work properly if newlines are possible, but it turns out we never sort when newlines are possible, because the results are generally nonsensical anyway (ever tried sorting the output of EXPLAIN)?

3. Otherwise, we assume there is no whitespace in result values. We expect the number of whitespace-delimited tokens in each line to equal the number of columns (and panic if that's not the case). All the sorting options can be used.

**Comment by knz** [comment]
Ok what do you think about  making our test result reformatter (`-rewrite-results`) behave as follows:

- if the test already uses `rowsort`, assume that the test implementor already knew what they were doing and were avoiding multi-line result cells, and do nothing. otherwise,
- run the query to produce a reference output, and wrap any multi-line results on a single line.
- suggest/add `rowsort` as necessary.

We could even choose a special character to represent newlines in result cells, and have the test check for newline characters in that way. In the past we've had success with the special unicode character "␤" (0x2424). Thoughts?

**Comment by andy-kimball** [comment]
I'm not opposed to it. One trouble is that if we format multi-line outputs on a single line, they won't be human-readable. There's also the `partialsort` cases to worry about, since they do not work with any whitespace in the result (not just newlines).

**Comment by yuzefovich** [comment]
Somewhat related idea was suggested by Radu in #27635 : it would be great if logic tests checker could verify whether a particular output is a possible result of a query. It would be beneficial for cases when SQL standard allows for multiple results (one example being order-dependent window function applications without `ORDER BY` clause).

**Comment by RaduBerinde** [comment]
@yuzefovich I think you misunderstood, I wasn't suggesting that, I was just saying that your change to fix tests is good. I'm not sure how the logic test could figure out that a certain result is possible.

The direction I would go here would be to automatically stress-test any modified logic tests in CI. In most cases, these flakes show up very quickly (handful of runs). We already do something that's similar in principle - we detect new go unit tests and run them under stress (and this has been useful).

**Comment by yuzefovich** [comment]
This has been addressed by @mgartner in #98626.
